gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeBuilder.java

```diff
@@ -120,27 +140,35 @@ private QueryNode buildExecutionTreeForSameSourceFilterAndSelection(String sourc
   }
 
   private QueryNode buildExecutionTreeForQsFilterAndSelection(String source) {
-    int selectionLimit = executionContext.getEntitiesRequest().getLimit();
-    int selectionOffset = executionContext.getEntitiesRequest().getOffset();
+    EntitiesRequest entitiesRequest = executionContext.getEntitiesRequest();
+    Filter filter = entitiesRequest.getFilter();
+    int selectionLimit = entitiesRequest.getLimit();
+    int selectionOffset = entitiesRequest.getOffset();
+    List<OrderByExpression> orderBys = entitiesRequest.getOrderByList();
 
     // query-service/Pinot does not support offset when group by is specified. Since we will be
     // grouping by at least the entity id, we will compute the non zero pagination ourselves. This
     // means that we need to request for offset + limit rows so that we can paginate appropriately.
     // Pinot will do the ordering for us.
+    // https://github.com/apache/incubator-pinot/issues/111#issuecomment-214810551
     if (selectionOffset > 0) {
       selectionLimit = selectionOffset + selectionLimit;
       selectionOffset = 0;
     }
 
-    QueryNode rootNode = new SelectionAndFilterNode(source, selectionLimit, selectionOffset);
+    QueryNode rootNode =
+        new DataFetcherNode(source, filter, selectionLimit, selectionOffset, orderBys);
+
     if (executionContext.getEntitiesRequest().getOffset() > 0) {
```

Comment:
nit: if (selectionOffset > 0)

Reply:
Can't do `if (selectionOffset > 0)`

There is a code block above which says

```
if (selectionOffset > 0) {
  selectionLimit = selectionOffset + selectionLimit;
  selectionOffset = 0;
}
```

Reply:
Ahh sorry.